doc/en/Configuration-file.md
# Configuration file

This file allows the user to customize the behavior of RHVoice. It is read-only
once at startup. So your changes will take effect only after you restart
the application which is using RHVoice.

On Windows, the configuration file is named `RHVoice.ini`. on other platforms it
is named `RHVoice.conf`.

## Path

### GNU/Linux

The path to the directory where the settings are stored is determined at compile
time. It can be redefined by setting the value of the `sysconfdir` variable when
you execute `scons`. By default, the value is `$prefix/etc`, which, if the
default value of the `prefix` variable is used, means `/usr/local/etc`.

The full path to the configuration file is `$sysconfdir/RHVoice/RHVoice.conf`.

### Windows

#### SAPI5

The path is `%APPDATA%\RHVoice\RHVoice.ini`.

#### NVDA

Open the folder where NVDA stores the user settings.
The `RHVoice.ini` file must be put into the `RHVoice-config` subfolder.

### Android

Depending on your device, the configuration file may be stored either in
internal memory or on an SD card. This will likely be somewhere in a subdirectory of `/storage`. 
The relative file path is
`Android/data/com.github.olga_yakovleva.rhvoice.android/files/RHVoice.conf`.
Users will often not have edit permissions to a file in this location. Instead, use the import functionality located in the app's settings, which will copy a file from a location you select to this location. This will overwrite any copy of the file previously placed in the aforementioned path.

## Format

RHVoice uses the classic ini format. But as this format is not defined by a
formal standard and implementations differ, here are some implementation details
users should keep in mind.

* The file must be saved in UTF-8.
* Semicolons are used to indicate comments. Only the whole line can be commented
  out.
* Periods are used as decimal separators.
* Characters in values may be represented with their numeric codes. The format
  is similar to numeric character references in XML, but the initial ampersand
  sign is omitted. For example, the English letter A can be written as `#65;`.
* If an option takes only two boolean values, on or off,  you can use the
  following keywords to enable the corresponding behavior: `true`, `yes`, `on`
  or `1`, and to disable it, use `false`, `no`, `off` or `0`.

## General and specific settings

Some of the options allow you to specify that they must be applied only to a
specific language or voice. And some of them are only defined for specific
languages.

### Language-specific settings

The following format is used:

```ini
languages.<language>.<key>=<value>
```

The language can be indicated by:

* Its English name
* its two-letter code defined by ISO 639-1
* its three-letter code defined by ISO 639-3.

Examples:

```ini
languages.russian.default_rate=2
languages.eng.default_rate=0.8
```

### Voice-specific settings

The following format is used:

```ini
voices.<name>.<key>=<value>
```

Example:

```ini
voices.elena.enabled=no
```

## Available options

### Rate, pitch and volume

The settings in this group determine how RHVoice interprets the values of rate,
pitch and volume it receives from client programs.

The value `1` corresponds to the standard behavior of the voices.

For example, NVDA expresses the value of speech rate as a percentage. RHVoice
interprets 50% as the default speech rate (`default_rate`), and 100% is
interpreted as the maximum speech rate (`max_rate`).

The following table lists all the options in this group.

| Setting          | Description    | Default value | Minimum value | Maximum value |
| ---------------- | -------------- | ------------- | ------------- | ------------- |
| `default_rate`   | default rate   | 1             | `min_rate`    | `max_rate`    |
| `min_rate`       | minimum rate   | 0.5           | 0.2           | 1             |
| `max_rate`       | maximum rate   | 2             | 1             | 5             |
| `default_pitch`  | default pitch  | 1             | `min_pitch`   | `max_pitch`   |
| `min_pitch`      | minimum pitch  | 0.5           | 0.5           | 1             |
| `max_pitch`      | maximum pitch  | 2             | 1             | 2             |
| `default_volume` | default volume | 1             | `min_volume`  | `max_volume`  |
| `min_volume`     | minimum volume | 0.25          | 0.25          | 1             |
| `max_volume`     | maximum volume | 2             | 1             | 4             |

These settings can be applied both to all the voices and to individual languages
or voices.

Examples:

```ini
default_volume=0.8
languages.english.default_rate=1.5
voices.alan.default_pitch=0.9
```

#### Using the Sonic library

RHVoice can use [the Sonic library](https://github.com/waywardgeek/sonic) for
changing speech rate. The native algorithm of speech rate modification in
RHVoice was improved in version 1.4.0, and this library is not included in the
builds provided by the RHVoice project. So the following setting is only
supported in custom builds with Sonic enabled.

The `min_sonic_rate` setting specifies the minimum rate
value starting from which Sonic will be used instead of the built-in algorithm.
By default the built-in algorithm is always used on Android, and on other
platforms Sonic is used to speed up speech, that is the synthesizer behaves as
if the user wrote in the configuration file:

```ini
min_sonic_rate=1
```

### Language switching and voice profiles

RHVoice can analyze the input text and change the synthesis language
automatically. This functionality is currently implemented only for those
languages whose alphabets share no letters. .

#### Voice profiles

A voice profile is just a combination of voices selected by the user. The
languages of the voices must differ. The first voice in the group is considered
the primary voice and determines the so-called primary language.

Screen readers and other client programs will include the voice profiles in the
list of available voices. But on Android voice profiles aren't configured via
this configuration file. .

A voice profile is written as a sequence of names of voices separated by the `+`
character. The `voice_profiles` setting is used to define the list of voice
profiles. Its value must be a comma-separated list of voice profiles.

Example:

```ini
voice_profiles=anna+slt,aleksandr+alan
```

#### Language switching

When one of the voice profiles is active, RHVoice will try to use the voice
which corresponds to the language of the text. The language may be set by the
client program or detected automatically. If the language cannot be detected
automatically, as is the case when reading individual numbers or names of
punctuationmarks, the primary voice will be used, that is the first voice of the
profile.

But if numbers or non-alphabetic characters appear in a sentence, and RHVoice
determines that the language of the sentence differs from the primary language,
there are two possibilities: read them with the primary voice or read the whole
sentence with one voice according to the language. By default the first
alternative is chosen, but the user can change this behavior by disabling the
`prefer_primary_language` setting:

```ini
prefer_primary_language=false
```

### Speech quality

The `quality` setting allows the user to choose one of the three available
speech quality options.  The lower the quality, the better the performance: most
importantly, the less time it will take for the synthesizer to start speaking.
The available modes are described in the following table..

| Value      | Sampling rate (kHz) | Optimized response time |
| ---------- | ------------------- | ----------------------- |
| `max`      | 24                  | no                      |
| `standard` | 24                  | yes                     |
| `min`      | 16                  | yes                     |

The standard quality is used by default, that is the synthesizer behaves as if
the user wrote in the configuration file:

```ini
quality=standard
```

### Punctuation

Despite the title, the settings in this group apply to other non-alphabetic
characters as well, even though, strictly speaking, they can't be considered
punctuation marks.

#### Punctuation mode

The `punctuation_mode` setting determines whether punctuation marks and other
symbols will be spoken. The following modes are supported:

| Value  | Description                      |
| ------ | -------------------------------- |
| `none` | don't read any symbols (default) |
| `some` | selective reading                |
| `all`  | read all symbols                 |

For example, reading of all punctuation is enabled this way:

```ini
punctuation_mode=all
```

#### Selective reading of punctuation

The `punctuation_list` setting is used to specify which symbols to speak when
the punctuation mode is set to `some`. For example:

```ini
punctuation_list=@$/\
```

#### Names of symbols

The built-in dictionary of punctuation marks and other symbols can't be called
comprehensive.  But users can define the names of additional symbols in the user
dictionary.

### Capital letters

The following settings determine whether and how the synthesizer will indicate
capital letters when it reads individual characters.

#### Capital letter indication mode

This mode is specified by the `indicate_capitals` setting. The following values
are supported:

| Value   | Description                   |
| ------- | ----------------------------- |
| `no`    | disabled (default)            |
| `word`  | say "capital" before a letter |
| `pitch` | change speech pitch           |
| `sound` | play a short sound            |

Example:

```ini
indicate_capitals=pitch
```

#### Changing pitch to indicate capital letters

If capital letters are indicated by a change in pitch, than the
`cap_pitch_factor` setting specifies how much to change the pitch in this
situation. This setting can be applied both to the synthesizer as a whole and to
individual voices.

The following example specifies a thirty percent increase:

```ini
cap_pitch_factor=1.3
```

### Stress marks

Stress indication in a text is supported only for Russian and Ukrainian
languages. For example, to tell the synthesizer that the vowel after the `+`
sign must receive stress, Add the following to the configuration file:

```ini
stress_marker=+
```

### Disabling individual languages and voices

You can disable some of the languages or voices. It may be useful if you don't
use them and don't want your screen reader to show them. The following example
disables all the English voices and Aleksandr::

```ini
languages.english.enabled=false
voices.aleksandr.enabled=false
```

### Pseudo-English mode

This mode is available only if the language doesn't use the Latin alphabet.

When they read English words, the voices can try to follow the correct English
pronunciation as much as possible. Though they can't use the correct English
sounds, but the closest sounds in their native language.

To use this mode, you need to install the English language pack. The English
voices don't need to be installed.

If the English language pack is installed, the pseudo-English mode is enabled
by
default. Using Russian as an example, you can disable it this way:

```ini
languages.russian.use_pseudo_english=no
```


doc/ru/Configuration-file.md
# файл конфигурации

Данный файл позволяет настроить некоторые аспекты поведения RHVoice. В текущей
версии файл считывается только один раз, при инициализации синтезатора.
Это означает, что для того, чтобы изменения настроек вступили в силу, необходимо
перезапустить программу, использующую RHVoice.

В Windows имя файла конфигурации - `RHVoice.ini`, на других платформах -
`RHVoice.conf`.

## Расположение

### GNU/Linux

Путь к директории с пользовательскими настройками определяется при компиляции.
Его можно переопределить, задав значение переменной `sysconfdir` при вызове
`scons`. По умолчанию используется значение `$prefix/etc`, что, при
использовании значения переменной `prefix` по умолчанию, означает
`/usr/local/etc`.

Полный путь к конфигурационному файлу - `$sysconfdir/RHVoice/RHVoice.conf`.

### Windows

#### Версия для SAPI5

Путь к файлу - `%APPDATA%\RHVoice\RHVoice.ini`.

#### Версия для NVDA

Перейдите в папку, где NVDA хранит все пользовательские настройки.
Файл `RHVoice.ini` должен находиться в подпапке `RHVoice-config`.

### Android

В зависимости от устройства, конфигурационный файл может располагаться либо во
внутренней памяти, либо на карте памяти. Путь к файлу -
`Android/data/com.github.olga_yakovleva.rhvoice.android/files/RHVoice.conf`.

## Формат

В целом RHVoice придерживается классического ini-формата. Поскольку этот формат
не задаётся никаким официальным стандартом, и существуют различия в реализации
между разными приложениями, далее перечислены некоторые особенности, на которые
следует обратить внимание.

* Файл должен сохраняться в кодировке UTF-8.
* В качестве символа комментария используется точка с запятой.
  Закомментировать можно только целую строку.
* В качестве десятичного разделителя используется точка.
* В значениях допускается представление любого символа его кодом. Формат
  представления соответствует формату числовых ссылок на символы,
  используемых в XML, только опускается начальный символ `&`.
  Например, русскую букву А можно записать следующим образом: `#1040;`.
* Если для опции допустимы только два логических значения, включена или
  отключена, то для включения можно использовать одно из ключевых слов:
  `true`, `yes`, `on` или `1`, а для отключения - `false`, `no`, `off` или `0`.

## Общие настройки и исключения

Для некоторых опций возможно указать, что они должны применяться только к
конкретному языку или голосу. Некоторые настройки доступны только для какого-то
одного языка.

### Настройки языков

Используется следующий формат:

```ini
languages.<язык>.<параметр>=<значение>
```

Язык можно задать одним из следующих способов:

* Английское название языка.
* Двухбуквенный код языка, согласно ISO 639-1.
* Трёхбуквенный код языка, согласно ISO 639-3.

Примеры:

```ini
languages.russian.default_rate=2
languages.eng.default_rate=0.8
```

### Настройки голосов

Используется следующий формат:

```ini
voices.<имя>.<параметр>=<значение>
```

Пример:

```ini
voices.elena.enabled=no
```

## Доступные опции

### Скорость, высота и громкость

Настройки из данной группы определяют, как RHVoice должен интерпретировать
значения скорости, высоты и громкости, которые он получает от программ-клиентов.

Значение `1` соответствует стандартному поведению голосов.

Например, NVDA передаёт синтезатору значения скорости, выраженные в процентах.
RHVoice интерпретирует 50% как скорость по умолчанию (`default_rate`),
а 100% - как максимальную скорость (`max_rate`).

В следующей таблице показаны все опции этой группы.

| Параметр         | Описание               | Значение по умолчанию | Минимальное значение | Максимальное значение |
| ---------------- | ---------------------- | --------------------- | -------------------- | --------------------- |
| `default_rate`   | скорость по умолчанию  | 1                     | `min_rate`           | `max_rate`            |
| `min_rate`       | минимальная скорость   | 0.5                   | 0.2                  | 1                     |
| `max_rate`       | максимальная скорость  | 2                     | 1                    | 5                     |
| `default_pitch`  | высота по умолчанию    | 1                     | `min_pitch`          | `max_pitch`           |
| `min_pitch`      | минимальная высота     | 0.5                   | 0.5                  | 1                     |
| `max_pitch`      | максимальная высота    | 2                     | 1                    | 2                     |
| `default_volume` | громкость по умолчанию | 1                     | `min_volume`         | `max_volume`          |
| `min_volume`     | минимальная громкость  | 0.25                  | 0.25                 | 1                     |
| `max_volume`     | максимальная громкость | 2                     | 1                    | 4                     |

Эти настройки можно задать как для синтезатора в целом, так и для отдельных
языков и голосов.

Примеры:

```ini
default_volume=0.8
languages.russian.default_rate=1.5
voices.aleksandr.default_pitch=0.9
```

#### Использование библиотеки Sonic

RHVoice может использовать [библиотеку
Sonic](https://github.com/waywardgeek/sonic) для изменения скорости
речи. Собственный алгоритм изменения скорости речи в RHVoice был
усовершенствован в версии 1.4.0, и библиотека Sonic больше не
включается в сборки, предоставляемые проектом RHVoice. Поэтому
следующая настройка поддерживается только в пользовательских сборках,
если было включено использование Sonic.

Настройка `min_sonic_rate` задаёт минимальное значение скорости, начиная с
которого вместо встроенного алгоритма изменения скорости будет использоваться
Sonic. По умолчанию в Android всегда используется встроенный алгоритм, а на
других платформах Sonic используется для ускорения речи, то есть синтезатор
ведёт себя так, как будто в конфигурационном файле прописано:

```ini
min_sonic_rate=1
```

### Переключение языка и голосовые профили

RHVoice может анализировать входной текст и автоматически переключать язык
синтеза. Эта функциональность доступна только для языков, чьи алфавиты не
включают общих букв.

#### Голосовые профили

Голосовой профиль - это просто выбранное пользователем сочетание голосов. Языки
этих голосов должны отличаться. Первый голос в группе считается основным голосом
и задаёт так называемый основной язык.

Чтецы экрана и другие программы-клиенты будут включать голосовые профили в
список доступных голосов. Исключение составляет Android, где, из-за особенностей
этой платформы, пользовательские голосовые профили не учитываются.

В конфигурационном файле голосовой профиль представляет собой последовательность
имён голосов, разделённых знаком `+`. Набор голосовых профилей задаётся
параметром `voice_profiles`. Его значение - список профилей, разделённых
запятой.

Пример:

```ini
voice_profiles=anna+slt,aleksandr+alan
```

#### Переключение языков

Когда активен один из голосовых профилей, RHVoice постарается использовать тот
голос, который соответствует языку текста. Язык может быть задан
программой-клиентом или определён автоматически. Если не удаётся определить
язык, как при чтении отдельных чисел или названий знаков пунктуации, будет
использован основной голос - первый голос профиля.

Если же такие элементы, как числа или неалфавитные символы, встречаются внутри
предложения, и RHVoice определил, что язык этого предложения отличается от
основного, возможны два варианта поведения синтезатора: переключаться на
основной голос при чтении этих элементов или читать всё предложение одним
голосом, соответствующим языку. По умолчанию используется первый вариант,
но пользователь может переопределить поведение синтезатора, отключив опцию
`prefer_primary_language`:

```ini
prefer_primary_language=false
```

### Качество речи

Настройка `quality` позволяет выбрать один из трёх доступных вариантов качества
речи.  Чем ниже качество, тем выше производительность: прежде всего, тем быстрее
синтезатор начнёт говорить. Доступные режимы описаны в следующей таблице.

| Значение   | Частота дискретизации (кГц) | Оптимизация времени отклика |
| ---------- | --------------------------- | --------------------------- |
| `max`      | 24                          | нет                         |
| `standard` | 24                          | да                          |
| `min`      | 16                          | да                          |

По умолчанию используется стандартное качество, то есть синтезатор ведёт себя
так, как будто в конфигурационном файле прописано:

```ini
quality=standard
```

### Чтение пунктуации

Несмотря на название, настройки этой группы относятся и к другим неалфавитным
символам, даже если их нельзя, строго говоря, классифицировать как знаки
препинания.

#### Режим чтения пунктуации

Этот режим задаётся параметром `punctuation_mode`. Доступны следующие варианты:

| Значение | Описание                          |
| -------- | --------------------------------- |
| `none`   | не читать (значение по умолчанию) |
| `some`   | выборочное чтение                 |
| `all`    | читать все знаки                  |

Например, включить чтение всех знаков препинания можно следующим образом:

```ini
punctuation_mode=all
```

#### Выборочное чтение пунктуации

Параметр `punctuation_list` задаёт, какие именно знаки произносить при
выборочном чтении. Например:

```ini
punctuation_list=@$/\
```

#### Названия знаков

Встроенный словарь знаков препинания и других символов нельзя назвать полным.
Но пользователь может задать названия дополнительных символов в пользовательском
словаре.

### Заглавные буквы

Следующие настройки определяют, будет ли синтезатор при чтении отдельных
символов уведомлять пользователя, что буква является заглавной.

#### Режим индикации заглавных букв

Этот режим задаётся параметром `indicate_capitals`. Поддерживаются следующие
значения:

| Значение | Описание                                   |
| -------- | ------------------------------------------ |
| `no`     | отключить (значение по умолчанию)          |
| `word`   | произносить слово "заглавная" перед буквой |
| `pitch`  | изменять высоту голоса                     |
| `sound`  | подавать короткий звуковой сигнал          |

Пример:

```ini
indicate_capitals=pitch
```

#### Изменение высоты при произнесении заглавной буквы

Если включено изменение высоты голоса при чтении заглавных букв, то параметр
`cap_pitch_factor` задаёт значение этого изменение. Эта настройка доступна как
для синтезатора в целом, так и для отдельных голосов.

Следующий пример устанавливает увеличение высоты на 30%:

```ini
cap_pitch_factor=1.3
```

### Указание ударения в тексте

Указание положения ударения в тексте поддерживается только для русского и
украинского языков. Например, чтобы указать синтезатору, что следующая за
знаком `+` гласная должна получить ударение, нужно прописать в конфигурационном
файле:

```ini
stress_marker=+
```

### Отключение отдельных языков и голосов

Вы можете сделать некоторые языки и голоса недоступными программам-клиентам.
Такая возможность может оказаться полезной, если вы не используете эти языки или
голоса и не хотите, чтобы ваш чтец экрана их отображал. Следующий пример
блокирует все английские голоса и Александра:

```ini
languages.english.enabled=false
voices.aleksandr.enabled=false
```

### Поддержка псевдо-английского

Данный режим доступен только для языков, не использующих латинский алфавит.

При чтении английских слов голоса могут стараться по мере возможности следовать
правильному английскому произношению. При этом они не могут использовать
правильные английские звуки, а только наиболее близкие звуки родного языка.

Чтобы можно было использовать этот режим, необходимо, чтобы был установлен
английский языковой пакет. При этом не обязательно устанавливать английские
голоса.

Поддержка псевдо-английского включена по умолчанию, если английский языковой
пакет установлен. Отключить её, например, для русских голосов можно так:

```ini
languages.russian.use_pseudo_english=no
```


NEWS
                        New in version 0.4-a2

A new Russian female voice has been included. It is trained on a
larger speech database, recorded by Irina Vorobyova. As usual, it
seems to have speech quality problems because of pitch extraction
failures, which is especially noticable in utterance final syllables.
I tried to model interrogative intonation explicitly by including many
questions in the recording script, but the attempt seems to be only
partially successful.

Several new configuration settings have been added, including
pseudo-english mode for Russian voices.

I now use the following two sources of information about stress
placement in Russian: a stress dictionary generated from RuLex, which
is a Russian pronunciation dictionary by Igor Poretsky, and a stress
dictionary extracted from Russian Wiktionary.

This version of RHVoice introduces a native implementation of an
output module for Speech Dispatcher. Though some features, such as reading of
key presses, have not been implemented yet.

                          New in version 0.4-a1

This is actually not an update, but a completely new synthesizer.
Being not the best of programmers in general and especially bad at
programming in pure C, I found it more and more difficult to maintain
that old code. Some new features were just impossible to implement
using the existing infrastructure. The most important of those changes
I couldn't make was multilingual support. So I rewrote the whole thing
from scratch in C++.

This version is not ready for production use. Some features are
incomplete, some are still missing, and there must be serious bugs I
haven't discovered yet.

At the moment RHVoice can speak in three languages: Russian, English
and Esperanto. There is only one Esperanto voice, and I expect it to
be the most problematic in terms of speech quality, because the
speaker often used creaky voice during the recording.

RHVoice now supports automatic language switching, if the languages
have no common letters. So at the moment there are two language pairs:
Russian+English and Russian+Esperanto. Whichever interface is used,
the client must set the main voice, which determines the primary
language. An optional extra voice will determine, if the language
switching should be enabled. If it's turned on and the synthesizer
cannot detect the language of the sentence, the primary language will
be used.

The concept of variants has been dropped for now, because it doesn't
really fit a multilingual text-to-speech system, and I am not sure
what I should do about it.

The previous version was painful to use from command line, because it
took a long time to load the voice in memory. The new one is even
worse in this respect, because language data is also stored in
external binary files and it takes time to load them. That's why i
haven't even written a proper command-line interface for the new
RHVoice yet. But I've implemented a D-Bus service and a command-line
client for it.

It is now possible to set language-specific and voice-specific prosody
options in the configuration file, e.g. you can use faster rate with
your native language. The sonic library is now the only supported
method of speeding up or slowin down speech. Support for applying
filters to speech is temporarily dropped, but I plan to return it in
0.5. Punctuation levels and capitalization indication haven't veen
implemented yet either, but they should be implemented before the
final release of 0.4.

User dictionaries will most likely come back only in 0.5, although
the format will be different (simpler, no regular expressions).

The worst news is probably that the C API is not backward compatible.
Since there will be at least three clients/interfaces developed by me
personally, I fixed what I think was wrong with the old API:
* There are no global initialization/uninitialization functions
  anymore, clients must create their own private instances of the tts engine.
* There won't be any global settings, even for the particular instance
  of the engine, because synthesis requests are usually processed in a
  separate thread, and global settings don't make sense in this case.
* There are now separate callbacks for different types of events, and
  most of them are optional.


src/core/params.cpp
/* Copyright (C) 2012, 2014, 2018, 2020  Olga Yakovleva <yakovleva.o.v@gmail.com> */

/* This program is free software: you can redistribute it and/or modify */
/* it under the terms of the GNU Lesser General Public License as published by */
/* the Free Software Foundation, either version 2.1 of the License, or */
/* (at your option) any later version. */

/* This program is distributed in the hope that it will be useful, */
/* but WITHOUT ANY WARRANTY; without even the implied warranty of */
/* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the */
/* GNU Lesser General Public License for more details. */

/* You should have received a copy of the GNU Lesser General Public License */
/* along with this program.  If not, see <https://www.gnu.org/licenses/>. */

#include "core/params.hpp"

namespace
{
  const double MAX_MAX_VOLUME=4;
  const double MAX_MAX_RATE=5;
  const double NO_MIN_SONIC_RATE=MAX_MAX_RATE+1;
}

#ifndef MAX_VOLUME
#define MAX_VOLUME 2
#endif

#ifndef MAX_RATE
#define MAX_RATE 2
#endif

#ifndef MIN_SONIC_RATE
#define MIN_SONIC_RATE 1
#endif

#ifndef DEFAULT_PUNCTUATION_MODE
#define DEFAULT_PUNCTUATION_MODE RHVoice_punctuation_none
#endif

namespace RHVoice
{
  voice_params::voice_params():
    default_rate("default_rate",1,0.2,MAX_MAX_RATE),
    min_rate("min_rate",0.5,0.2,1),
    max_rate("max_rate",MAX_RATE,1,MAX_MAX_RATE),
    default_pitch("default_pitch",1,0.5,2),
    min_pitch("min_pitch",0.5,0.5,1),
    max_pitch("max_pitch",2,1,2),
    default_volume("default_volume",1,0.25,MAX_MAX_VOLUME),
    min_volume("min_volume",0.25,0.25,1),
    max_volume("max_volume",MAX_VOLUME,1,MAX_MAX_VOLUME),
    cap_pitch_factor("cap_pitch_factor",1.3,0.5,2),
    min_sonic_rate("min_sonic_rate",MIN_SONIC_RATE,0.2,NO_MIN_SONIC_RATE)
  {
  }

  void voice_params::default_to(const voice_params& other)
  {
    default_rate.default_to(other.default_rate);
    min_rate.default_to(other.min_rate);
    max_rate.default_to(other.max_rate);
    default_pitch.default_to(other.default_pitch);
    min_pitch.default_to(other.min_pitch);
    max_pitch.default_to(other.max_pitch);
    default_volume.default_to(other.default_volume);
    min_volume.default_to(other.min_volume);
    max_volume.default_to(other.max_volume);
    cap_pitch_factor.default_to(other.cap_pitch_factor);
    min_sonic_rate.default_to(other.min_sonic_rate);
  }

  void voice_params::register_self(config& cfg,const std::string& prefix)
  {
    cfg.register_setting(default_rate,prefix);
    cfg.register_setting(min_rate,prefix);
    cfg.register_setting(max_rate,prefix);
    cfg.register_setting(default_pitch,prefix);
    cfg.register_setting(min_pitch,prefix);
    cfg.register_setting(max_pitch,prefix);
    cfg.register_setting(default_volume,prefix);
    cfg.register_setting(min_volume,prefix);
    cfg.register_setting(max_volume,prefix);
    cfg.register_setting(cap_pitch_factor,prefix);
    cfg.register_setting(min_sonic_rate,prefix);
  }

  text_params::text_params():
    stress_marker("stress_marker",'+')
  {
  }

  void text_params::default_to(const text_params& other)
  {
    stress_marker.default_to(other.stress_marker);
  }

  void text_params::register_self(config& cfg,const std::string& prefix)
  {
    cfg.register_setting(stress_marker,prefix);
  }

  absolute_speech_params::absolute_speech_params():
    rate("rate",0,-1,1),
    pitch("pitch",0,-1,1),
    volume("volume",0,-1,1)
  {
  }

  void absolute_speech_params::default_to(const absolute_speech_params& other)
  {
    rate.default_to(other.rate);
    pitch.default_to(other.pitch);
    volume.default_to(other.volume);
  }

  relative_speech_params::relative_speech_params():
    rate("rate",1,0.1,10),
    pitch("pitch",1,0.5,2),
    volume("volume",1,0.25,2)
  {
  }

  void relative_speech_params::default_to(const relative_speech_params& other)
  {
    rate.default_to(other.rate);
    pitch.default_to(other.pitch);
    volume.default_to(other.volume);
  }

  verbosity_params::verbosity_params():
    punctuation_mode("punctuation_mode",DEFAULT_PUNCTUATION_MODE),
    punctuation_list("punctuation_list","+=<>~@#$%^&*|"),
    capitals_mode("indicate_capitals",RHVoice_capitals_off)
  {
    punctuation_mode.define("none",RHVoice_punctuation_none);
    punctuation_mode.define("some",RHVoice_punctuation_some);
    punctuation_mode.define("all",RHVoice_punctuation_all);
    capitals_mode.define("off",RHVoice_capitals_off);
    capitals_mode.define("no",RHVoice_capitals_off);
    capitals_mode.define("none",RHVoice_capitals_off);
    capitals_mode.define("false",RHVoice_capitals_off);
    capitals_mode.define("word",RHVoice_capitals_word);
    capitals_mode.define("pitch",RHVoice_capitals_pitch);
    capitals_mode.define("sound",RHVoice_capitals_sound);
  }

  void verbosity_params::default_to(const verbosity_params& other)
  {
    punctuation_mode.default_to(other.punctuation_mode);
    punctuation_list.default_to(other.punctuation_list);
    capitals_mode.default_to(other.capitals_mode);
  }

  void verbosity_params::register_self(config& cfg,const std::string& prefix)
  {
    cfg.register_setting(punctuation_mode,prefix);
    cfg.register_setting(punctuation_list,prefix);
    cfg.register_setting(capitals_mode,prefix);
  }

  void stream_params::register_self(config& cfg)
  {
    cfg.register_setting(fixed_size);
    cfg.register_setting(view_size);
  }
}
